airbyte-config/config-models/src/main/java/io/airbyte/config/helpers/StateMessageHelper.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.config.helpers;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.StateType;
+import io.airbyte.config.StateWrapper;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import java.util.List;
+import java.util.Optional;
+
+public class StateMessageHelper {
+
+  public static class AirbyteStateMessageListTypeReference extends TypeReference<List<AirbyteStateMessage>> {}
+
+  /**
+   * This a takes a json blob state and tries return either a legacy state in the format of a json
+   * object or a state message with the new format which is a list of airbyte state message.
+   *
+   * @param state
+   * @return Either a json blob (on the left) or a structure state message.
+   */
+  public static Optional<StateWrapper> getTypedState(JsonNode state) {
+    if (state == null) {
+      return Optional.empty();
+    } else {
+      try {
+        List<AirbyteStateMessage> stateMessages = Jsons.object(state, new AirbyteStateMessageListTypeReference());
+        if (stateMessages.size() == 1 && stateMessages.get(0).getStateType() == AirbyteStateType.GLOBAL) {
+          return Optional.of(new StateWrapper()
+              .withStateType(StateType.GLOBAL)
+              .withGlobal(stateMessages.get(0)));
+        } else if (stateMessages.size() >= 1
+            && stateMessages.stream().filter(stateMessage -> stateMessage.getStateType() != AirbyteStateType.STREAM).toList().isEmpty()) {
```

Comment:
why do you need this `>= 1 check`? is `stateMessages.stream().allMatch(stateMessage -> stateMessage.getStateType() == AirbyteStateType.STREAM)` enough?

Reply:
FWIW, I do something similar to this in the `StateManagerFactory` by just doing looking at the first element and then switching from there.  The assumption is that all state messages will have the same type.  This may be a faulty assumption and could be remedied with a precondition check that makes sure that all messages have the same type.  This would remove all of the size/index checking going on in this code.  Something like:

```
final Optional<AirbyteStateMessage> airbyteStateMessage = stateMessages.stream().findFirst();
if (airbyteStateMessage.isPresent()) {
    switch(airbyteStateMessage.get().getStateType()) {
        case GLOBAL:
            return Optional.of(new StateWrapper() ...)
        case STREAM:
            return Optional(... )
        case LEGACY:
        default:
            return Optional(...)
    }
} else {
    // Handle empty state list?
}
```
Something like that.